Fixes #10727 - Added validation to require BMC Smart proxy

ShimShtein opened the pull request:
in order to add BMC nic.

domcleal commented:
This could stop people from being able to store BMC info in Foreman without using its BMC functionality, no?  I think Discovery may populate BMC interfaces too, which doesn't imply you have the proxy capability.

ShimShtein commented:
Added overrides in case the NIC is not managed

domcleal commented on app/models/nic/bmc.rb:
```diff
@@ -39,6 +37,24 @@ def ensure_physical
+
+    private
+
+    def get_proxy
```
Nitpick, I'd call this `bmc_proxy` to be specific and for consistency with smart proxy associations on other models.

domcleal commented on app/models/nic/bmc.rb:
```diff
@@ -39,6 +37,24 @@ def ensure_physical
+      return true unless managed?
+      return true if host && !host_managed?
+      unless get_proxy
+        errors.add(:base, 'There is no proxy with feature ''BMC'' set up. Please register a smart proxy with this feature.')
```
This string needs extracting, and avoid using `''` inside as gettext probably won't extract it properly.

domcleal commented on test/unit/nics/bmc_test.rb:
```diff
@@ -5,4 +5,39 @@ class BMCTest < ActiveSupport::TestCase
+    test 'requires BMC proxy in the same subnet' do
+      host = FactoryGirl.build(:host, :managed)
+      bmc_nic = FactoryGirl.build(:nic_bmc, :with_subnet, :host => host)
+      bmc_nic.subnet.proxies.stubs(:find).returns([])
```
Does this need stubbing, if it's a new subnet?

domcleal replied:
> About the issue with the stubbing, it is created with a default smart proxy. I prefer to be sure that the test will not break, so I set up the proxy.

You mean the default association to dhcp/dns in the subnet factory?  Better to just assign them to nil I think so it's not created.

unorthodoxgeek replied:
please use rails 4 compatible syntax
bmc_nic.subnet.proxies.where(whatever).all
and the compatible syntax in the test.

unorthodoxgeek replied:
@domcleal - I strongly prefer stubbing over touching actual data in a test, even if it's just in memory data.

domcleal replied:
Ok, I don't.  Setting up data X, stubbing (not even expecting) it to act a little like Y (but not completely) isn't as precise as having the factory set up Y in the first place.

ares replied:
I don't see a benefit in stubbing in this case either (also because stubbing AR query methods is more than tricky), if the test builds records using factory girl, it can build exactly those it needs for the test

domcleal commented on test/factories/host_related.rb:
```diff
@@ -55,6 +55,10 @@ def set_nic_attributes(host, attributes, evaluator)
     username 'admin'
     password 'admin'
+
+    trait :with_subnet do
```
I'd move this into the base factory, it's not BMC-specific.

domcleal commented on test/unit/nics/bmc_test.rb:
```diff
@@ -5,4 +5,39 @@ class BMCTest < ActiveSupport::TestCase
+      host = FactoryGirl.build(:host, :managed)
+      bmc_nic = FactoryGirl.build(:nic_bmc, :host => host)
+      host.interfaces << bmc_nic
+      refute host.valid?
```
Test bmc_nic's validity rather than host, and do you even need `host`?

domcleal replied:
Ditto on other tests about validity, try to test just the NIC object rather than all of host.  Even better if you can test that this particular error has been added via assert_valid/refute_with_errors or something.

dLobatog commented:
@ShimShtein I'm thinking the same thing @domcleal mentioned, at most show a more informative message when it's not found but we should still allow to save BMC info even without BMC proxies. The issue reported originally doesn't seem like a bug to me except for the Foreman::Exception being thrown.

ShimShtein commented:
@eLobato I am not enforcing proxy, if the NIC is not managed - it should be enough for inventory-only use cases.
@domcleal : I have re-checked the host issue that you have mentioned, existence of host is enforced. Although I have changed the tests to check the validity of the NIC, and not the host. About the issue with the stubbing, it is created with a default smart proxy. I prefer to be sure that the test will not break, so I set up the proxy.
I have addressed all other issues, thanks.

domcleal commented on test/unit/nics/bmc_test.rb:
```diff
@@ -5,4 +5,39 @@ class BMCTest < ActiveSupport::TestCase
+      host = FactoryGirl.build(:host, :managed => false)
+      bmc_nic = FactoryGirl.build(:nic_bmc, :host => host)
+      host.interfaces << bmc_nic
+      assert host.valid?
```
also bmc_nic

domcleal commented:
When you change the contents of a PR, please reply to or address the comments I raise, otherwise it's hard to tell when or what has changed.

ShimShtein commented:
@domcleal @ares And now without stubbing

domcleal commented on test/unit/nics/bmc_test.rb:
```diff
@@ -5,4 +5,45 @@ class BMCTest < ActiveSupport::TestCase
+      host = FactoryGirl.build(:host, :managed)
+      subnet = FactoryGirl.build(
+        :subnet,
+        :tftp => nil,
```
It looks like the default associations of :tftp and :dhcp should be removed from the factory, which would allow this and other tests simply to use a sensible default of an empty subnet sans associations.

The only thing it breaks is `DhcpOrchestrationTest#test_0006_provision interface DHCP records should contain filename/next-server attributes` which tries to create a subnet via the host factory with both, which won't work (as one trait overwrites the other), so we could move the slight ugliness to that test instead, and improve the factory defaults.

domcleal commented on app/models/nic/bmc.rb:
```diff
@@ -39,6 +37,24 @@ def ensure_physical
+      return true unless managed?
+      return true if host && !host_managed?
+      unless bmc_proxy
+        errors.add(:base, N_('There is no proxy with BMC feature set up. Please register a smart proxy with this feature.'))
```
I don't see this anywhere in the UI, are base errors actually displayed for interfaces?

ShimShtein commented:
Done with the subnet cleanup. We still have smart proxy fixtures, so I had to leave the `destroy_all` in the setup.

Moved the error to the `type` field, so now you can see it on the UI.

domcleal commented:
@ShimShtein thanks, looks fine.

@ares do you agree with this?  My only reservation is that it links whether the NIC's managed (i.e. if we create records) to whether it actually functions as a BMC interface for Foreman's purposes, which is a slightly different things.

dLobatog commented:
I don't think conflating "managed", which is already a charged term in Foreman (managed host vs managed NIC, for instance), with BMC is right, it's a different thing. In my opinion the bug described in Redmine can be fixed by handling the error better.

ares commented:
We already have this issue with other parts like orchestration. If NIC and it's host is managed we create DHCP record and also we create config file for this interface. Similarly we run some validations only if interface is managed. While I agree managed flag is user for too many things, to keep it consistent I'm OK with this (actually I'd expect this behaviour).

dLobatog commented:
:ok: then, but handle this case which is in the end why the issue was open:

`Moreover, when I see the BMC tab on host details page and on clicking the 'BMC' tab, it was showing: Failure: ERF42-1518 [Foreman::Exception]: Unable to find a proxy with BMC feature`

Validating the interface will not matter much anyway if you remove the proxy/or the proxy is down/etc.. , it will show an exception on that Host#show tab.

ShimShtein commented:
@eLobato, After this change, the error you have mentioned should not be raised unless you actually delete relevant Smart Proxies from Foreman. It does not handle runtime problems like the proxy is down etc.
To handle the deletion case, we probably need to add validation on the smart proxy object. I don't think it's so common, but if I'm wrong I can add it as a part of this PR.

domcleal commented on app/models/nic/bmc.rb:
```diff
@@ -39,6 +37,24 @@ def ensure_physical
+      if subnet.present?
+        proxy = subnet.proxies.find { |subnet_proxy| subnet_proxy.has_feature?('BMC') }
+      end
+      proxy ||= SmartProxy.with_features("BMC").first
```
Default scoping will apply taxonomies, so this could prevent saving of a host if a smart proxy isn't visible to one particular user.  Should this be unscoped, or scoped specifically to the same as the host?

domcleal commented:
I don't think that preventing deletion of smart proxies in this instance is a very good idea, it could be problematic when there's no hard association between a NIC and the smart proxy.  Usually we only enforce hard associations.  This is a bit of an issue with the design of this lookup method to begin with.

ShimShtein committed:
Fixes #10727 - Added validation to require BMC Smart proxy in order to add BMC nic

ShimShtein commented:
@domcleal I have unscoped the proxies - there is no need that the proxy will be in the same organization/location as the host. The user doesn't need to know about that proxy too. As long as the subnet/domain are exposed to the user, he should be able to create a host with a BMC interface on those networks.

About deletion - I am against it too. I think it would cause more trouble than good.

domcleal commented on app/models/nic/bmc.rb:
```diff
@@ -39,6 +37,24 @@ def ensure_physical
+
+    def bmc_proxy
+      if subnet.present?
+        proxy = subnet.proxies.find { |subnet_proxy| subnet_proxy.has_feature?('BMC') }
```
Taxonomy (default) scoping might apply to `subnet.proxies` too?

ShimShtein replied:
@domcleal `subnet.proxies` is a simple array, it's not an `ActiveRecord::Relation` look [here](https://github.com/theforeman/foreman/blob/develop/app/models/subnet.rb#L196)

domcleal replied:
It was the underlying associations I was a bit worried about, but in my tests it seems that the default taxonomy scope of smart proxy doesn't seem to apply to `subnet.dns` etc., so I guess that's OK.

domcleal commented:
Merged as a9c70fb, thanks @ShimShtein.
